3.0/m_relaymsg.cpp

```diff
@@ -0,0 +1,209 @@
+            return CMD_FAILURE;
+        }
+
+        // Check that target nick is not already in use
+        if (ServerInstance->FindNick(nick))
+        {
+            user->WriteNumeric(ERR_BADRELAYNICK, nick, "RELAYMSG spoofed nick is already in use");
+            return CMD_FAILURE;
+        }
+
+        // Make sure the nick doesn't include any core IRC characters (e.g. *, !)
+        // This should still be more flexible than regular nick checking - in particular
+        // we want to allow "/" and "~" for relayers
+        if (nick.find_first_of("!+%@&#$:'\"?*,.") != std::string::npos)
+        {
+            user->WriteNumeric(ERR_BADRELAYNICK, nick, "Invalid characters in spoofed nick");
+            return CMD_FAILURE;
+        }
+
+        // If the sender was a lcoal user, check that the target includes a nick separator
+        if (IS_LOCAL(user))
+        {
+            if (nick.find_first_of(cap.nick_separators) == std::string::npos)
+            {
+                user->WriteNumeric(ERR_BADRELAYNICK, nick, InspIRCd::Format("Spoofed nickname must include one of the following separators: %s", cap.nick_separators.c_str()));
+                return CMD_FAILURE;
+            }
+        }
+
+        // Send the message to everyone in the channel
+        std::string fakeSource = GetFakeHostmask(nick);
+        ClientProtocol::Messages::Privmsg privmsg(fakeSource, channel, text);
+        // Tag the message as @draft/relaymsg=<nick> so the sender can recognize it
+        // Also copy over tags set on the original /relaymsg command
+        privmsg.AddTags(parameters.GetTags());
+        privmsg.AddTag("draft/relaymsg", &captag, user->nick);
+        channel->Write(ServerInstance->GetRFCEvents().privmsg, privmsg);
+
+        if (IS_LOCAL(user))
+        {
+            // Pass the message on to other servers
+            CommandBase::Params params;
+            // Preserve other message tags sent with the /relaymsg command
+            params.GetTags().insert(parameters.GetTags().begin(), parameters.GetTags().end());
+            params.push_back(channame);
+            params.push_back(nick);
+            params.push_back(":" + text);
+
+            ServerInstance->PI->BroadcastEncap("RELAYMSG", params, user);
+        }
+
+        return CMD_SUCCESS;
+    };
+};
+
+class ModuleRelayMsg : public Module
+{
+    RelayMsgCap cap;
+    RelayMsgCapTag captag;
+    CommandRelayMsg cmd;
+
+public:
+    ModuleRelayMsg() :
+        cap(this),
+        captag(this, cap),
+        cmd(this, cap, captag)
+    {
+    }
+
+    void ReadConfig(ConfigStatus& status) CXX11_OVERRIDE
+    {
+        ConfigTag* tag = ServerInstance->Config->ConfValue("relaymsg");
+        cap.nick_separators = tag->getString("separators", "/", 1);
+        cmd.fake_ident = tag->getString("ident", "relay");
+        cmd.fake_host = tag->getString("host", ServerInstance->Config->ServerName);
+
+        if (!ServerInstance->IsIdent(cmd.fake_ident))
```

Comment:
A possible issue: if a user edits their config and changes from a valid ident/host to an invalid ident/host then even if this throws the malformed config will sitll be applied.

To work around this you can put fake_ident and fake_host into a local variable then swap after all of the checks have been performed (also, move nick_separators to the end so it doesn't get applied till after the checks too).